### PR TITLE
mgr/dashboard: redesign the login screen

### PR DESCRIPTION
**Pull request by ishanrai05**
fixes : https://tracker.ceph.com/issues/35689
Signed-off-by: Ishan Rai <ishanrai05@gmail.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

## Checklist
- [ *] References tracker ticket
- [ *] Updates documentation if necessary
- [ *] Includes tests for new functionality or reproducer for bug

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ishanrai05** [comment]
ss

**Comment by bk201** [comment]
Thanks for your contribution!
Please also add the `Fixes: xxxx` line in commit message.

**Comment by ishanrai05** [comment]
Do we want the background colour of focus to change? If yes then what colour?

**Comment by votdev** [comment]
@Exotelis Can you please have a look if this PR if it is OK related to branding.

File: src/pybind/mgr/dashboard/frontend/src/app/core/layouts/login-layout/login-layout.component.html [hunk 1]
```diff
@@ -10,13 +10,11 @@
...
-          <img src="assets/Ceph_Logo_Stacked_RGB_White_120411_fa_256x256.png"
+        <div class="col-sm-12 col-md-6 d-sm-block">
+          <img src="assets/Ceph_Logo_Standard_RGB_White_120411_fa.png"
                alt="Ceph"
-               class="float-right img-fluid">
-        </div>
-        <div class="col-12 col-sm-6 col-xl-5">
-          <router-outlet></router-outlet>
+               class="img-fluid mb-5">
+               <router-outlet></router-outlet>
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/layouts/login-layout/login-layout.component.html]: Indention is incorrect.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html [hunk 2]
```diff
@@ -1,41 +1,51 @@
...
-               required>
-        <span class="input-group-append">
-          <button type="button"
-                  class="btn btn-outline-light btn-password"
+    <div class="form-group has-feedback" id="password-div">
+      <div class="input-group d-flex flex-nowrap">
+        <div class="d-flex flex-column flex-grow-1 py-3">
+          <label class="placeholder pl-4" for="password" i18n>Password</label>
+          <input id="password"
+                  name="password"
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html]: Indention is incorrect

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html [hunk 3]
```diff
@@ -1,41 +1,51 @@
...
+                  placeholder=""
+                  class="form-control pl-4"
+                  required
+                  id="password"
+                  autofocus onfocus = "document.getElementById('password-div').classList.add('color');" 
+                  onfocusout = "document.getElementById('password-div').classList.remove('color');">
+        </div>
+        <span class="form-group-append">
+        <button type="button"
+                  class="btn btn-outline-light btn-password h-100 px-4"
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html]: Indention is incorrect.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html [hunk 4]
```diff
@@ -1,41 +1,51 @@
...
-             type="text"
-             placeholder="Enter your username..."
-             class="form-control"
-             required
-             autofocus>
+    <div class="form-group has-feedback" id="username-div" for="username">
+      <div class="d-flex flex-column py-3">
+        <label class="placeholder pl-4" for="username" i18n>Username</label>
+        <input name="username"
+              [(ngModel)]="model.username"
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html]: Indention is incorrect.

**Comment by ishanrai05** [comment]
I am getting a test failure. I am not exactly sure what the error means.
```bash
Summary of all failing tests
 FAIL  src/app/ceph/pool/pool-details/pool-details.component.spec.ts (25.923s)
  ● PoolDetailsComponent › Pool details tabset › pool data filtering is called

    HttpErrorResponse: Http failure response for api/summary: 0 Unknown Error

      87 |     });
      88 | 
    > 89 |     it('pool data filtering is called', () => {
         |     ^
      90 |       const filterNonPoolDataSpy = spyOn(
      91 |         poolDetailsComponent,
      92 |         'filterNonPoolData'

      at new Spec (node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
      at src/app/ceph/pool/pool-details/pool-details.component.spec.ts:89:5
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:391:26)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone.js:150:43)
      at src/app/ceph/pool/pool-details/pool-details.component.spec.ts:36:3
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:391:26)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone.js:150:43)
      at Object.<anonymous> (src/app/ceph/pool/pool-details/pool-details.component.spec.ts:12:1)
```

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html [hunk 5]
```diff
@@ -1,41 +1,57 @@
...
+                 i18n>Password</label>
+          <input id="password"
+                 name="password"
+                 [(ngModel)]="model.password"
+                 #password="ngModel"
+                 type="password"
+                 placeholder=""
+                 class="form-control pl-4"
+                 required
+                 autofocus onfocus="document.getElementById('password-div').classList.add('color');"
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html]:
```suggestion
                 autofocus 
                 onfocus="document.getElementById('password-div').classList.add('color');"
```

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html [hunk 6]
```diff
@@ -1,41 +1,57 @@
...
+               i18n>Username</label>
+        <input id="username"
+               name="username"
+               [(ngModel)]="model.username"
+               #username="ngModel"
+               type="text"
+               placeholder=""
+               class="form-control pl-4"
+               required
+               autofocus onfocus="document.getElementById('username-div').classList.add('color');"
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html]:
```suggestion
               autofocus 
               onfocus="document.getElementById('username-div').classList.add('color');"
```

**Comment by callithea** [comment]
> ```shell
> it('pool data filtering is called', () => {
> ```

Hi @ishanrai05, this is a random error tracked [here](https://tracker.ceph.com/issues/42929) and is not related to your PR - just re-trigger make check in case you face this issue again.

**Comment by callithea** [comment]
> I am getting a test failure. I am not exactly sure what the error at line 89 means.
> 
> ```shell
> Summary of all failing tests
>  FAIL  src/app/ceph/pool/pool-details/pool-details.component.spec.ts (25.923s)
>   ● PoolDetailsComponent › Pool details tabset › pool data filtering is called
> 
>     HttpErrorResponse: Http failure response for api/summary: 0 Unknown Error
> 
>       87 |     });
>       88 | 
>     > 89 |     it('pool data filtering is called', () => {
>          |     ^
>       90 |       const filterNonPoolDataSpy = spyOn(
>       91 |         poolDetailsComponent,
>       92 |         'filterNonPoolData'
> 
>       at new Spec (node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
>       at src/app/ceph/pool/pool-details/pool-details.component.spec.ts:89:5
>       at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:391:26)
>       at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone.js:150:43)
>       at src/app/ceph/pool/pool-details/pool-details.component.spec.ts:36:3
>       at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:391:26)
>       at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone.js:150:43)
>       at Object.<anonymous> (src/app/ceph/pool/pool-details/pool-details.component.spec.ts:12:1)
> ```

Hi @ishanrai05, this is a random error tracked [here](https://tracker.ceph.com/issues/42929) and is not related to your PR - just re-trigger make check in case you face this issue again.

**Comment by votdev** [comment]
Can you please modify the PNG file and make the background transparent? I see differences in the background color of the image and the HTML page.

Bildschirmfoto von 2020-03-26 11-59-18

Or you need to change `#374249` to `#3a444a`.

**Comment by ishanrai05** [comment]
sss
@votdev  Hi. I am getting a `#3374249` on the image but I believe we should make the background transparent anyway. Although I am not very sure how to do it?

**Comment by ishanrai05** [comment]
@callithea Actually I was getting make check errors on Jenkins as well, so I thought I must have broken some tests. Thank you. I will check again.

**Comment by votdev** [comment]
> sss
> @votdev Hi. I am getting a `#3374249` on the image but I believe we should make the background transparent anyway. Although I am not very sure how to make the background of png image transparent.

The render issue appears in Firefox because of the color profile embedded in the PNG file.

See https://css-tricks.com/color-rendering-difference-firefox-vs-safari/ or https://stackoverflow.com/questions/35820231/css-background-color-and-png-color-inconsistent-between-browsers for example.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.scss [hunk 7]
```diff
@@ -5,15 +5,31 @@
...
+    border-left: 4px solid white;
+    background-color: $color-password-toggle-bg;
+    height: auto;
+    margin-bottom: 2rem;
+  }
+
+  #username-div.color,
+  #password-div.color {
+    border-left: 4px solid $color-accent;
+  }
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.scss:18]: Just found out about this. You can use :focus-within to check if any child component is focused.
With this you don't need to keep changing classes in the html.

.invalid-feedback is not working well with this type of forms, add this style to fix it.

```suggestion
    &:focus-within {
      border-left: 4px solid $color-accent;
    }
  }

  .invalid-feedback {
    display: block;
  }
```
I would suggest adding more left padding to the error messages:
image

**Comment by ishanrai05** [comment]
I'll update the PR. Can we also do something about the background image? It is not being rendered properly on firefox.

**Comment by votdev** [comment]
Maybe we should try a different image format (e.g. GIF, JPEG or SVG) to workaround the Firefox ICC color profile issue (https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/3.5/ICC_color_correction_in_Firefox).

**Comment by epuertat** [comment]
Same issue happens to the (already merged) about box (and almost every color looks different in Firefox vs. Chrome):

Chrome:
image

Firefox:
image

What about using the first 1-pixel width as a sprite and filling the rest of the login page with that ([[reference]](https://stackoverflow.com/questions/15087412/put-background-image-in-center-and-repeat-last-pixel-to-left-and-right)).

**Comment by ishanrai05** [comment]
@votdev  gif seems to work for me. jpeg is worse and image is not being converted properly to svg.

@epuertat I'll update the about box after this.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html [hunk 8]
```diff
@@ -1,41 +1,59 @@
...
+               name="username"
+               [(ngModel)]="model.username"
+               #username="ngModel"
+               type="text"
+               placeholder=""
+               class="form-control pl-4"
+               required
+               autofocus
+               onfocus="document.getElementById('username-div').classList.add('color');"
+               onfocusout="document.getElementById('username-div').classList.remove('color');">
```
[the remark this answers is not in the document]
Comment (ishanrai05) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.html]: I made the changes.

**Comment by ishanrai05** [comment]
@epuertat Hi. The sprite method does not work on my chrome. It works perfectly fine on firefox. I think it is not supported by all browsers. I upgraded the gif file.

**Comment by ishanrai05** [comment]
@epuertat Is the test failure related to the PR?

**Comment by tspmelo** [comment]
jenkins test make check

**Comment by epuertat** [comment]
@votdev could you please check if all your changes requested have been addressed, and if so approve it? Thanks!

**Pull request merged**